File: Day 2/solution.py
from re import search

f = open('input.txt')
fin = [i.strip() for i in f.readlines()]


def part_one():
    valid = 0
    for line in fin:
        data = search(r'^(\d+)-(\d+) ([a-z]): ([a-z]+)$', line)

        rng = range(int(data.group(1)), int(data.group(2)) + 1)
        lt = data.group(3)
        password = data.group(4)

        if password.count(lt) in rng:
            valid += 1
            
    return valid


def part_two():
    valid = 0
    for line in fin:
        data = search('(\d+)-(\d+) ([a-z]): ([a-z]+)', line)

        a, b = int(data.group(1)), int(data.group(2))
        lt = data.group(3)
        password = data.group(4)
        
        if (password[a-1] == lt) ^ (password[b-1] == lt):
            valid += 1
            
    return valid


if __name__ == '__main__':
    print("Part 1:", part_one())
    print("Part 2:", part_two())








        

File: Day 2/solution.rs
use regex::Regex;

fn main() {
    println!("Part 1: {}", part_one());
    println!("Part 2: {}", part_two())
}

fn load_input() -> Vec<String> {
    let file = std::fs::read_to_string("input.txt")
        .expect("Couldn't open the file");

    file
        .lines()
        .map(|s| s.to_owned())
        .collect::<Vec<String>>()
}

fn part_one() -> i32 {
    let inp = load_input();
    let re = Regex::new(r"(\d+)-(\d+) ([a-z]): ([a-z]+)").unwrap();

    let mut count = 0;
    for line in inp {
        let caps = re.captures(&line).expect("Can't find any match");

        let a: usize = caps[1].parse().unwrap();
        let b: usize = caps[2].parse().unwrap();
        let lt = &caps[3];
        let password = &caps[4];
        
        if (a..=b).contains(&(password.matches(&lt).count())) {
            count += 1;
        }
    }
    count
}

fn part_two() -> i32 {
    let inp = load_input();
    let re = Regex::new(r"(\d+)-(\d+) ([a-z]): ([a-z]+)").unwrap();

    let mut count = 0;
    for line in inp {
        let caps = re.captures(&line).expect("Can't find any match");

        let a = caps[1].parse::<usize>().unwrap() - 1;
        let b = caps[2].parse::<usize>().unwrap() - 1;
        let lt = &caps[3];
        let password = &caps[4];
        
        if (&password[a..=a] == lt) ^ (&password[b..=b] == lt) {
            count += 1;
        }
    }
    count
}